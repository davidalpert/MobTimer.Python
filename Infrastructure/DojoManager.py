import _thread
import random
import uuid
import paho.mqtt.publish as mqtt_pub
import paho.mqtt.client as mqtt
import json

TOPIC_MOBBER_LIST = "MobberList"

TOPIC_TIME_CHANGE = "TimeChange"

TOPIC_SAY_HELLO = "SayHello"

TOPIC_START_TIMER = "StartTimer"


class DojoManager(object):
    def __init__(self, controller):
        if not controller.settings_manager.get_dojo_enabled():
            return

        self.time_change_increment = 0
        self.mobber_change_increment = 0
        self.session_id = uuid.uuid4().__str__()

        self.switch_dictionary = {
            TOPIC_MOBBER_LIST: self.sub_mobber_list,
            TOPIC_TIME_CHANGE: self.sub_time_change,
            TOPIC_SAY_HELLO: self.sub_say_hello,
            TOPIC_START_TIMER: self.sub_start_timer,
        }
        self.station_drivers = {}
        self.other_stations = []
        self.controller = controller
        self.dojo_broker = self.controller.settings_manager.get_dojo_broker()
        self.dojo_port = self.controller.settings_manager.get_dojo_port()
        self.dojo_station_uuid = uuid.uuid4().__str__()
        self.dojo_mob_station_name = self.controller.settings_manager.get_dojo_mob_station_name()
        self.dojo_session_id = self.controller.settings_manager.get_dojo_session_id()
        self.dojo_topic_root = self.controller.settings_manager.get_dojo_topic_root()
        self.controller.mobber_manager.subscribe_to_mobber_list_change(self.publish_mobber_list_changes)
        self.controller.time_options_manager.subscribe_to_timechange(self.publish_time_change)
        self.controller.countdown_manager.subscribe_to_time_changes(self.publish_countdown_change)
        self.subscribe_to_time_change()
        self.say_hello()

    def say_hello(self):
        topic = self.generate_topic(TOPIC_SAY_HELLO)
        self.publish(topic, "")

    def publish_countdown_change(self, negative_if_time_expired, minutes, seconds, origin_station_name=None):
        if negative_if_time_expired < 0: return
        print("publish_countdown_change" , minutes, seconds)
        topic = self.generate_topic(TOPIC_START_TIMER)
        self.publish(topic, "")

    def publish_time_change(self, time_string, minutes, seconds, origin_station_name=None):
        topic = self.generate_topic(TOPIC_TIME_CHANGE)
        station_name = self.dojo_mob_station_name
        station_uuid = self.dojo_station_uuid
        if origin_station_name is not None:
            return
        payload_dictionary = {
            "minutes": minutes,
            "seconds": seconds,
            "station_name": station_name,
            "station_uuid": station_uuid
        }
        payload = json.dumps(payload_dictionary)
        self.publish(topic, payload)

    def publish_mobber_list_changes(self, mobber_list, driver_index, next_driver_index, navigator_index):
        topic = self.generate_topic(TOPIC_MOBBER_LIST)
        payload_object = {
            "driver_index": driver_index,
            "next_driver_index": next_driver_index,
            "navigator_index": navigator_index,
            "mobber_list": mobber_list
        }
        payload = json.dumps(payload_object)
        self.publish(topic, payload)

    def publish(self, topic, payload):
        _thread.start_new_thread(self.thread_publish, (topic, payload))

    def thread_publish(self, topic, payload):
        mqtt_pub.single(topic, hostname=self.dojo_broker, port=self.dojo_port, payload=payload)

    def subscribe_to_time_change(self):
        _thread.start_new_thread(self.subscribe_to_dojo, ())

    def on_connect(self, client, userdata, flags, rc):
        topic = "{}/{}/#".format(self.dojo_topic_root, self.dojo_session_id)
        client.subscribe(topic)

    def on_message(self, client, userdata, msg):
        topic_parts = msg.topic.split('/')
        topic_root = topic_parts[0]
        session_id = topic_parts[1]
        station_uuid = topic_parts[2]
        station_name = topic_parts[3]
        message_type = topic_parts[4]
        print("on_message", msg.topic, "station_uuid", station_uuid)
        self.switch_statement_dictionary_trick(station_uuid, station_name, message_type, msg.payload)

    def switch_statement_dictionary_trick(self, station_uuid, station_name, message_type, payload):
        self.switch_dictionary[message_type](station_uuid, station_name, message_type, payload)

    def sub_mobber_list(self, station_uuid, station_name, message_type, payload):
        if not station_uuid == self.dojo_station_uuid:
            payload_dictionary = json.loads(payload.decode("utf-8"))
            mobber_list = payload_dictionary["mobber_list"]
            print("sub_mobber_list", mobber_list)
            self.controller.mobber_manager.set_mobber_list(mobber_list)

    def generate_topic(self, message_type):
        topic = "{}/{}/{}/{}/{}".format(self.dojo_topic_root, self.dojo_session_id, self.dojo_station_uuid,
                                        self.dojo_mob_station_name,
                                        message_type)
        print("generate_topic", topic)
        return topic

    def sub_time_change(self, station_uuid, station_name, message_type, payload):
        print(payload)
        if not station_uuid == self.dojo_station_uuid:
            payload_dictionary = json.loads(payload.decode("utf-8"))
            minutes = payload_dictionary["minutes"]
            seconds = payload_dictionary["seconds"]
            origin_station_name = payload_dictionary["station_name"]
            if not (
                            self.controller.time_options_manager.minutes == minutes and self.controller.time_options_manager.seconds == seconds):
                self.controller.time_options_manager.set_countdown_time(minutes, seconds, origin_station_name)

    def sub_say_hello(self, station_uuid, station_name, message_type, payload):
        if not station_uuid == self.dojo_station_uuid:
            if not self.other_stations.__contains__(station_name):
                self.other_stations.append(station_name)
                topic = self.generate_topic(TOPIC_SAY_HELLO)
                self.publish(topic, "")

    def sub_start_timer(self, station_uuid, station_name, message_type, payload):
        if not station_uuid == self.dojo_station_uuid:
            self.controller.launch_transparent_countdown_if_blocking()

    def subscribe_to_dojo(self):
        client = mqtt.Client()
        client.on_connect = self.on_connect
        client.on_message = self.on_message
        client.connect(self.dojo_broker, self.dojo_port, 60)
        client.loop_forever()
