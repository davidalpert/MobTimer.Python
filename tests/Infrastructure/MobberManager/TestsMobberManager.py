import os
import random
import unittest
from approvaltests import Approvals
from approvaltests.TextDiffReporter import TextDiffReporter
from Infrastructure.MobberManager import MobberManager

os.environ["APPROVALS_TEXT_DIFF_TOOL"] = "C:\\Program Files\\TortoiseSVN\\bin\\TortoiseMerge.exe"


class TestsMobberManager(unittest.TestCase):
    def test_empty_mobber_manager_has_no_items(self):
        mobber_manager = MobberManager()
        self.assertEqual(mobber_manager.mobber_count(), 0)

    def test_add_mobber_chris_has_chris(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Chris")
        result = ["Chris"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_mobber_joe_chris_has_joe_chris(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        result = ["Joe", "Chris"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_mobber_joe_chris_joe__remove_joe_has_joe_chris(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("John")
        mobber_manager.remove_mobber(2)
        result = ["Joe", "Chris"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_4_mobbers_move_up_middle(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Will")
        mobber_manager.add_mobber("Eric")
        mobber_manager.move_mobber_up(2)
        result = ["Joe", "Will", "Chris", "Eric"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_4_mobbers_move_up_top(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Will")
        mobber_manager.add_mobber("Eric")
        mobber_manager.move_mobber_up(0)
        result = ["Eric", "Chris", "Will", "Joe"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_4_mobbers_move_down_middle(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Will")
        mobber_manager.add_mobber("Eric")
        mobber_manager.move_mobber_down(2)
        result = ["Joe", "Chris", "Eric", "Will"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_add_4_mobbers_move_down_bottom(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Will")
        mobber_manager.add_mobber("Eric")
        mobber_manager.move_mobber_down(3)
        result = ["Eric", "Chris", "Will", "Joe"]
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_move_down_empty(self):
        mobber_manager = MobberManager()
        mobber_manager.move_mobber_down(0)
        result = []
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_move_up_empty(self):
        mobber_manager = MobberManager()
        mobber_manager.move_mobber_up(0)
        result = []
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_remove_empty(self):
        mobber_manager = MobberManager()
        mobber_manager.remove_mobber(0)
        result = []
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_clear(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Sam")
        mobber_manager.clear()
        result = []
        self.assertEqual(mobber_manager.get_mobbers(), result)

    def test_subscribe_to_mobber_list_changes(self):
        mobber_manager = MobberManager()
        result = {"result": "Mobbers in List for Each Change\n", "increment": 0}

        def time_change_callback(mobber_list, driver_index, next_driver_index, navigator_index):
            result["increment"] += 1
            result["result"] += "Action " + result["increment"].__str__() + ":"
            for mobber_index in range(0, mobber_list.__len__()):
                result["result"] += mobber_list[mobber_index]
                if mobber_index == driver_index:
                    result["result"] += " (Current)"
                if mobber_index == next_driver_index:
                    result["result"] += " (Next)"
                if mobber_index == navigator_index:
                    result["result"] += " (Navigator)"
                result["result"] += ", "

            result["result"] += "\n"

        mobber_manager.subscribe_to_mobber_list_change(time_change_callback)

        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Sam")
        mobber_manager.add_mobber("John")
        mobber_manager.switch_next_driver()
        mobber_manager.add_mobber("Bill")
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.remove_mobber(2)
        mobber_manager.remove_mobber(0)
        mobber_manager.switch_next_driver()
        mobber_manager.rewind_driver()
        mobber_manager.add_mobber("Seth")
        mobber_manager.rewind_driver()
        mobber_manager.rewind_driver()
        mobber_manager.rewind_driver()
        mobber_manager.move_mobber_down(0)
        mobber_manager.add_mobber("Fredrick")
        mobber_manager.move_mobber_up(2)
        mobber_manager.remove_mobber(1)
        mobber_manager.remove_mobber(0)
        mobber_manager.remove_mobber(0)

        Approvals.verify(result["result"], TextDiffReporter())

    def test_subscribe_to_mobber_list_changes_when_navigating_after_driving(self):
        mobber_manager = MobberManager(drive_after_navigating=False)
        result = {"result": "Mobbers in List for Each Change\n", "increment": 0}

        def time_change_callback(mobber_list, driver_index, next_driver_index, navigator_index):
            result["increment"] += 1
            result["result"] += "Action " + result["increment"].__str__() + ":"
            for mobber_index in range(0, mobber_list.__len__()):
                result["result"] += mobber_list[mobber_index]
                if mobber_index == driver_index:
                    result["result"] += " (Current)"
                if mobber_index == next_driver_index:
                    result["result"] += " (Next)"
                if mobber_index == navigator_index:
                    result["result"] += " (Navigator)"
                result["result"] += ", "

            result["result"] += "\n"

        mobber_manager.subscribe_to_mobber_list_change(time_change_callback)

        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Sam")
        mobber_manager.add_mobber("John")
        mobber_manager.switch_next_driver()
        mobber_manager.add_mobber("Bill")
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.remove_mobber(2)
        mobber_manager.remove_mobber(0)
        mobber_manager.switch_next_driver()
        mobber_manager.rewind_driver()
        mobber_manager.add_mobber("Seth")
        mobber_manager.rewind_driver()
        mobber_manager.rewind_driver()
        mobber_manager.rewind_driver()
        mobber_manager.move_mobber_down(0)
        mobber_manager.add_mobber("Fredrick")
        mobber_manager.move_mobber_up(2)
        mobber_manager.remove_mobber(1)
        mobber_manager.remove_mobber(0)
        mobber_manager.remove_mobber(0)

        Approvals.verify(result["result"], TextDiffReporter())

    def test_subscribe_to_mobber_list_changes_random(self):
        random.seed(0)
        mobber_manager = MobberManager(True)
        result = {"result": "Mobbers in List for Each Change\n", "increment": 0}

        def time_change_callback(mobber_list, driver_index, next_driver_index, navigator_index):
            result["increment"] += 1
            result["result"] += "Action " + result["increment"].__str__() + ":"
            for mobber_index in range(0, mobber_list.__len__()):
                result["result"] += mobber_list[mobber_index]
                if mobber_index == driver_index:
                    result["result"] += " (Current)"
                if mobber_index == next_driver_index:
                    result["result"] += " (Next)"
                if mobber_index == navigator_index:
                    result["result"] += " (Navigator)"
                result["result"] += ", "

            result["result"] += "\n"

        mobber_manager.subscribe_to_mobber_list_change(time_change_callback)

        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Sam")
        mobber_manager.add_mobber("John")
        mobber_manager.switch_next_driver()
        mobber_manager.add_mobber("Bill")
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.set_mobber_list(["Hello", "Eric", "Joe"])
        mobber_manager.switch_next_driver()
        mobber_manager.switch_next_driver()
        mobber_manager.remove_mobber(2)
        mobber_manager.remove_mobber(0)
        mobber_manager.switch_next_driver()
        mobber_manager.add_mobber("Seth")
        mobber_manager.move_mobber_down(0)
        mobber_manager.add_mobber("Fredrick")
        mobber_manager.move_mobber_up(2)
        mobber_manager.remove_mobber(1)
        mobber_manager.remove_mobber(0)
        mobber_manager.remove_mobber(0)

        Approvals.verify(result["result"], TextDiffReporter())

    def test_next_driver1_driver0_index(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        result = "Next: " + str(mobber_manager.next_driver_index) + " Current: " + str(mobber_manager.current_driver_index)
        self.assertEqual(result, "Next: 1 Current: 0")

    def test_switch_next_driver0_driver1_index(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.switch_next_driver()
        result = "Next: " + str(mobber_manager.next_driver_index) + " Current: " + str(mobber_manager.current_driver_index)
        self.assertEqual(result, "Next: 0 Current: 1")

    def test_navigator_defaults_to_next_driver(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        result = "Next: " + str(mobber_manager.next_driver_index) + " Current: " + str(mobber_manager.current_driver_index) + " Navigator: " + str(mobber_manager.navigator_index)
        self.assertEqual(result, "Next: 1 Current: 0 Navigator: 1")

    def test_navigator_moves_with_the_next_driver(self):
        mobber_manager = MobberManager()
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.switch_next_driver()
        result = "Next: " + str(mobber_manager.next_driver_index) + " Current: " + str(mobber_manager.current_driver_index) + " Navigator: " + str(mobber_manager.navigator_index)
        self.assertEqual(result, "Next: 0 Current: 1 Navigator: 0")

    def test_navigator_defaults_when_configuration_is_set(self):
        mobber_manager = MobberManager(drive_after_navigating=False)
        mobber_manager.add_mobber("Joe")
        mobber_manager.add_mobber("Chris")
        mobber_manager.add_mobber("Tracy")
        result = "Next: " + str(mobber_manager.next_driver_index) + " Current: " + str(mobber_manager.current_driver_index) + " Navigator: " + str(mobber_manager.navigator_index)
        self.assertEqual(result, "Next: 1 Current: 0 Navigator: 2")

if __name__ == '__main__':
    unittest.main()
