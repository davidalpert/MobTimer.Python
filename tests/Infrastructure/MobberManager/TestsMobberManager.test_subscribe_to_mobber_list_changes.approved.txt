Mobbers in List for Each Change
Action 1:
Action 2:Joe (Driver) (Navigator), 
Action 3:Joe (Driver), Chris (Navigator), 
Action 4:Joe (Driver), Chris (Navigator), Sam, 
Action 5:Joe (Driver), Chris (Navigator), Sam, John, 
Action 6:Joe, Chris (Driver), Sam (Navigator), John, 
Action 7:Joe, Chris (Driver), Sam (Navigator), John, Bill, 
Action 8:Joe, Chris, Sam (Driver), John (Navigator), Bill, 
Action 9:Joe, Chris, Sam, John (Driver), Bill (Navigator), 
Action 10:Joe (Navigator), Chris, Sam, John, Bill (Driver), 
Action 11:Joe (Driver), Chris (Navigator), Sam, John, Bill, 
Action 12:Joe, Chris (Driver), Sam (Navigator), John, Bill, 
Action 13:Joe, Chris (Driver), John (Navigator), Bill, 
Action 14:Chris, John (Driver), Bill (Navigator), 
Action 15:Chris (Navigator), John, Bill (Driver), 
Action 16:Chris, John (Driver), Bill (Navigator), 
Action 17:Chris, John (Driver), Bill (Navigator), Seth, 
Action 18:Chris (Driver), John (Navigator), Bill, Seth, 
Action 19:Chris (Navigator), John, Bill, Seth (Driver), 
Action 20:Chris, John, Bill (Driver), Seth (Navigator), 
Action 21:John, Chris, Bill (Driver), Seth (Navigator), 
Action 22:John, Chris, Bill (Driver), Seth (Navigator), Fredrick, 
Action 23:John, Bill, Chris (Driver), Seth (Navigator), Fredrick, 
Action 24:John, Chris, Seth (Driver), Fredrick (Navigator), 
Action 25:Chris (Navigator), Seth, Fredrick (Driver), 
Action 26:Seth (Driver), Fredrick (Navigator), 
